### ISSUE TITLE
[Mac M1 MPS] Incorrect matrix multiplication results for n-by-n matrices

### ISSUE THREAD

**Issue by chienhsiang** [pytorch/pytorch#84736]
### 🐛 Describe the bug

When multiplying two n-by-n matrices using either torch.mm, torch.matmul or @, the results are different in MPS device from the CPU. It seems that MPS automatically transposes the 2nd matrix. Is this an expected behavior?

~~~python
import torch

n = 3
a = torch.randn(n, n)
b = torch.randn(n, n)

device = torch.device('mps')
ag = a.to(device)
bg = b.to(device)

print("CPU results:")
print("> torch.mm")
print(torch.mm(a, b))
print()

print("> torch.matmul")
print(torch.matmul(a, b))
print()

print("> a @ b")
print(a @ b)
print()

print("GPU results:")
print("> torch.mm")
print(torch.mm(ag, bg))
print()

print("> torch.matmul")
print(torch.matmul(ag, bg))
print()

print("> ag @ bg")
print(ag @ bg)
print()

print("GPU results by transposing the 2nd matrix:")
print("> torch.mm")
print(torch.mm(ag, bg.T))
print()

print("> torch.matmul")
print(torch.matmul(ag, bg.T))
print()

print("> ag @ bg.T")
print(ag @ bg.T)
print()
~~~

Output:

~~~console
CPU results:
> torch.mm
tensor([[-2.0771, -1.6845,  1.6261],
        [-0.2620, -0.3063, -1.1082],
        [ 1.0023,  1.0229,  1.2402]])

> torch.matmul
tensor([[-2.0771, -1.6845,  1.6261],
        [-0.2620, -0.3063, -1.1082],
        [ 1.0023,  1.0229,  1.2402]])

> a @ b
tensor([[-2.0771, -1.6845,  1.6261],
        [-0.2620, -0.3063, -1.1082],
        [ 1.0023,  1.0229,  1.2402]])

GPU results:
> torch.mm
/Users/osiris/miniconda3/envs/ml/lib/python3.10/site-packages/torch/_tensor_str.py:103: UserWarning: The operator 'aten::bitwise_and.Tensor_out' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1659484612588/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
tensor([[-4.2804,  2.4117, -5.2009],
        [ 0.9807,  0.5783,  0.7883],
        [-0.2656, -1.5243,  0.3112]], device='mps:0')

> torch.matmul
tensor([[-4.2804,  2.4117, -5.2009],
        [ 0.9807,  0.5783,  0.7883],
        [-0.2656, -1.5243,  0.3112]], device='mps:0')

> ag @ bg
tensor([[-4.2804,  2.4117, -5.2009],
        [ 0.9807,  0.5783,  0.7883],
        [-0.2656, -1.5243,  0.3112]], device='mps:0')

GPU results by transposing the 2nd matrix:
> torch.mm
tensor([[-2.0771, -1.6845,  1.6261],
        [-0.2620, -0.3063, -1.1082],
        [ 1.0023,  1.0229,  1.2402]], device='mps:0')

> torch.matmul
tensor([[-2.0771, -1.6845,  1.6261],
        [-0.2620, -0.3063, -1.1082],
        [ 1.0023,  1.0229,  1.2402]], device='mps:0')

> ag @ bg.T
tensor([[-2.0771, -1.6845,  1.6261],
        [-0.2620, -0.3063, -1.1082],
        [ 1.0023,  1.0229,  1.2402]], device='mps:0')
~~~

### Versions

PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] numpy                     1.21.5          py310h220015d_3
[conda] numpy-base                1.21.5          py310h742c864_3
[conda] pytorch                   1.12.1                 py3.10_0    pytorch
[conda] torchaudio                0.12.1                py310_cpu    pytorch
[conda] torchvision               0.13.1                py310_cpu    pytorch

cc @kulinseth @albanD

**Comment by DenisVieriu97** [comment]
Hi @chienhsiang - thanks for the report! 
This issue should be fixed in PyTorch nightly - could you please give a try with latest nightly build and let me know if you still see the issue?

**Comment by chienhsiang** [comment]
Hi @DenisVieriu97,
Thank you for the response! Yes, the issue is fixed in pytorch=1.13.0.dev20220908.
